### ISSUE TITLE
Scylla startup never completes after populating counter tables and restarting

### ISSUE THREAD

**Issue by AndrewKlopper** [scylladb/scylla#2623]
*Installation details*
Scylla version (or git commit hash): 1.7.2-0.20170706.a24dcf1
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): Debian 8.8

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker): Linode VM
Hardware: cores=6 memory=8GB
Disks: (SSD/HDD, count): SSD x 1

We have a 3 node cluster that we are using to evaluate Scylla. We use counter tables with large numbers of counter columns.

We can create the cluster, create our keyspace with tables, populate the tables and query them without any issues, but if we attempt to restart Scylla on any of the nodes, the startup never completes even though the scylla process still appears to be running (systemd claims the service has failed to start, and the scylla listen ports are not opened).

Syslog shows no error or warning messages, and the last few entries refer to setting the compaction strategy for the system tables. There is no reference to the keyspace that we have created, so presumably scylla is getting stuck when it tries to open this keyspace.

A 'ps' shows the following scylla process still running:

/usr/bin/scylla --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --collectd-address=127.0.0.1:25826 --collectd=1 --collectd-poll-period 3000 --network-stack posix --max-io-requests=150

The only way for me to get scylla to start normally again appears to be deleting everything under /var/lib/scylla, creating the commitlog, coredump and data directories by hand, and starting again.

I have attached some syslogs in the hope that they shed some light on the matter, however I don't know how to proceed to debug this further: [syslog.gz](https://github.com/scylladb/scylla/files/1168322/syslog.gz)

**Comment by tgrabiec** [comment]
@AndrewKlopper It would be helpful to get a core dump of Scylla when it hangs. Try reproducing with `-c1 -m1G` to reduce core dump size. You can trigger the core dump with ```kill SIGABRT -p `pgrep scylla` ``` (try twice). 

http://docs.scylladb.com/operating-scylla/troubleshooting/report_scylla_problem/#core-dump

**Comment by AndrewKlopper** [comment]
I have uploaded the coredump to “scylladb-users-upload.s3.amazonaws.com/b85b3fcb-e3f0-45f0-8bb4-b20b09c66e82/core.scylla-2017-07-24-11:00:55-7380.xz”

 

Let me know if you need anything else.

**Comment by tgrabiec** [comment]
@AndrewKlopper How long did you wait since boot before taking the coredump?

**Comment by AndrewKlopper** [comment]
I waited about 20 minutes, but it seems systemd times out and kills/restarts the scylla process every 15 minutes, so according to the logs that particular coredump was taken 3 minutes and 10 seconds after the scylla process was started for the second time.

I have now started the server by hand and will leave it running for a few hours before doing the coredump, if that will help. So far, after 10 minutes it hasn't written any additional log entries beyond the ones I initially posted.

**Comment by AndrewKlopper** [comment]
I have uploaded another core to "scylladb-users-upload.s3.amazonaws.com/b85b3fcb-e3f0-45f0-8bb4-b20b09c66e82/core.scylla-2017-07-26-11:19:57-12310.xz". This one was taken after 18 hours. The scylla server still did not seem to have progressed any further in the startup.

**Comment by tgrabiec** [comment]
@AndrewKlopper Could you upload your system tables from `/var/lib/scylla/data` ?

**Comment by tgrabiec** [comment]
The second core dump was taken with `1.7.3-0.20170718.41b4055`.

**Comment by tgrabiec** [comment]
Can you try booting with the following line in `scylla.yaml` and/or bumping up scylla memory:

```
max_cached_partition_size_in_kb: 100000
```

This bumps up the default about ~10 times.

Below is why I think it would help.

The problem seems to be similar in both core dumps. Schema tables parsing is blocked on reading from `system.schema_columns`. There are 71 live readers. There are 61 readers waiting on `restricted_mutation_reader_config` semaphore (which limits read parallelism to 10) and there are 10 readers which are active.

I checked one of the 10 active readers. It belongs to a combined reader which is blocked on (eventually) `restricting_mutation_reader::operator()()::{lambda()#1}`, meaning it's blocked due to `restricting_mutation_reader` waiting to get a permit. It's not the active reader which tries to get the permit. The combined reader contains only two readers, the active one (cache) and an empty_reader (memtable). The active reader belongs to `single_partition_populating_reader`. The only way that could result in creation of another reader is if the partition was wide enough and it created the second non-populating reader. The `_uncached_wide_partitions` counter shows `10`, so this checks out. The stall is due to a deadlock on read permits caused by the fact that cache reader tries to create two readers at the same time.

@AndrewKlopper  What's your schema?

**Comment by AndrewKlopper** [comment]
I have uploaded my entire /var/lib/scylla directory including config and schema files (schema1.cql and schema2.cql) to "scylladb-users-upload.s3.amazonaws.com/b85b3fcb-e3f0-45f0-8bb4-b20b09c66e82/var_lib_scylla.tar.bz2". It contains only a small amount of test data. The individual tables do, however, have large numbers of counter columns, which is presumably what it triggering the issue.

Adding
```
max_cached_partition_size_in_kb: 100000
```
to scylla.yaml does indeed allow startup to continue, but it exposes another problem.

When I start up the first node, it blocks for 300 seconds due to an inability to connect to the other seeds. Unfortunately the other seeds block as well due to the same error. Eventually they all time out and the cluster completes starting up. The relevant log messages seem to be:
```
Jul 27 10:51:19 dmo-cass1-uk scylla:  [shard 0] gossip - Connect seeds again ... (299 seconds passed)
Jul 27 10:51:20 dmo-cass1-uk scylla:  [shard 0] storage_service - Checking remote features with gossip failed, fallback to check with system table
```
After that, everything seems to work fine.

**Comment by tgrabiec** [comment]
@AndrewKlopper I will send a proper patch for the deadlock problem soon (bumping up `max_cached_partition_size_in_kb` is just a workaround). Please open a new issue for the connection issues.

**Comment by AndrewKlopper** [comment]
There already seems to be an issue for the connection issues: #1679

**Comment by tgrabiec** [comment]
Fixed by 0ac2c388b6d3e6d36747cbe8088ed794bfa13a6d
